### PR TITLE
Make DTA _check_compatible_with less strict by default

### PR DESCRIPTION
**Pull request by jbrockmendel**
One of the two non-cosmetic things mentioned in #30720.

There are a bunch of places where DTA or DTI do a compatibility check that for tz_awareness_compat, but not requiring the same tz.  This check is analogous to `PeriodArray._check_compatible_with` and `TimedeltaArray._check_compatible_with`, so this adds a kwarg to _check_compatible_with so that we can use _check_compatible_with in all the relevant places and subsequently de-duplicate a bunch of code.

In addition to the comparisons, this is going to be relevant for searchsorted and insert, where we have slightly different behavior in a bunch of EA/Index subclasses.

### CODE REVIEW

File: pandas/core/arrays/datetimelike.py [hunk 1]
```diff
@@ -123,6 +123,9 @@ def _check_compatible_with(
         Parameters
         ----------
         other
+        strict : bool, default False
+            For __setitem__ we may have stricter compatiblity resrictions than
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: don’t we always need to check this for date times? why do u actually need a keyword

Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: ATM we dont use _check_compatible_with in the DTA comparison ops, but after this we will be able to.

Comment (jreback) [reply pandas/core/arrays/datetimelike.py]: can you call this something else besides strict as this only applies to DTI, maybe check_tz (and ignore this elsewhere). i know strict is more generic, but also very confusing.

Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: changed "strict"->"setitem"

**Comment by jbrockmendel** [comment]
updated+green

**Pull request merged**